src/uct/ib/dc/dc_mlx5.c

```diff
@@ -1349,37 +1321,46 @@ uct_dc_mlx5_dci_keepalive_handle_failure(uct_dc_mlx5_iface_t *iface,
     }
 
     ep = ucs_derived_of(op->ep, uct_dc_mlx5_ep_t);
-    if (ep->dci != UCT_DC_MLX5_EP_NO_DCI) {
+
+    if (ep->dci == UCT_DC_MLX5_EP_NO_DCI) {
+        ucs_assert(ep != iface->tx.fc_ep);
+        uct_dc_mlx5_iface_set_ep_failed(iface, ep, cqe, txwq, ep_status);
+    } else {
+        /* ep has another dci assinged to post operations, which sould be
+         * restarted too */
         uct_dc_mlx5_ep_handle_failure(ep, cqe, ep_status);
-        goto put_op;
     }
 
-    ucs_assert(ep != iface->tx.fc_ep);
-    uct_dc_mlx5_iface_set_ep_failed(iface, ep, cqe, txwq, ep_status);
-
 put_op:
     ucs_mpool_put(op);
 
-restart_dci:
-    uct_dc_mlx5_iface_restart_dci(iface, dci, ep_status);
+reset_dci:
+    uct_dc_mlx5_iface_reset_dci(iface, dci, ep_status);
 }
 
 ucs_status_t uct_dc_mlx5_iface_keepalive_init(uct_dc_mlx5_iface_t *iface)
 {
+    ucs_status_t status;
+
     if (ucs_likely(iface->flags & UCT_DC_MLX5_IFACE_FLAG_KEEPALIVE)) {
         return UCS_OK;
     }
 
     ucs_assert(iface->tx.ndci <= UCT_DC_MLX5_IFACE_MAX_USER_DCIS);
 
-    iface->flags |= UCT_DC_MLX5_IFACE_FLAG_KEEPALIVE;
-    return uct_dc_mlx5_iface_create_dci(iface, &iface->tx.dcis[iface->tx.ndci]);
+    status = uct_dc_mlx5_iface_create_dci(iface, &iface->tx.dcis[iface->tx.ndci]);
+    if (status == UCS_OK) {
+        iface->flags |= UCT_DC_MLX5_IFACE_FLAG_KEEPALIVE;
+    }
+
+    return status;
```

Comment:
for consistency with other codes:
status = uct_dc_ ...
if (status != UCS_OK) {
    return status;
}

iface->flags |= UCT_DC_MLX5_IFACE_FLAG_KEEPALIVE;
return UCS_OK;

Reply:
updated